oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -97,33 +119,157 @@ private static String encodeCert(X509Certificate certificate)
+            throw new IllegalArgumentException(
+                "The leaf certificate should only appear at the beginning of the trust chain file, or be omitted entirely.");
+          }
 
-      return OAuth2Utils.JSON_FACTORY.toString(certChain);
+          // Add the current certificate to the chain.
+          certChain.add(encodedCurrentCert);
+        }
+      }
+    } catch (IllegalArgumentException e) {
+      // This catches the specific error for misconfigured trust chain (e.g., leaf in wrong place).
+      throw new IOException("Trust chain misconfiguration: " + e.getMessage(), e);
+    } catch (NoSuchFileException e) {
+      throw new IOException(String.format("Trust chain file not found: %s", trustChainPath), e);
     } catch (CertificateException e) {
-      // Catch CertificateException to provide a more specific error message including
-      // the path of the file that failed to parse, and re-throw as IOException
-      // as expected by the getSubjectToken method signature for I/O related issues.
+      // This catches parsing errors if certificates in the trust chain file are invalid.
       throw new IOException(
-          "Failed to parse certificate(s) from: " + credentialSource.getCredentialLocation(), e);
+          String.format("Failed to parse certificate(s) from trust chain file: %s", trustChainPath),
+          e);
+    } catch (IOException e) {
+      // This catches any other general I/O errors during trust chain file reading (e.g.,
+      // permissions).
+      throw new IOException(
+          String.format("Failed to read trust chain file: %s", trustChainPath), e);
+    }
+
+    return OAuth2Utils.JSON_FACTORY.toString(certChain);
+  }
+
+  /**
+   * Reads a file containing PEM-encoded X509 certificates and returns a list of parsed
+   * certificates. It splits the file content based on PEM headers and parses each certificate.
+   * Returns an empty list if the trust chain path is empty.
+   *
+   * @param trustChainPath The path to the trust chain file.
+   * @return A list of parsed X509 certificates.
+   * @throws IOException If an error occurs while reading the file.
+   * @throws CertificateException If an error occurs while parsing a certificate.
+   */
+  @VisibleForTesting
+  static List<X509Certificate> readTrustChain(String trustChainPath)
+      throws IOException, CertificateException {
+    List<X509Certificate> certificateTrustChain = new ArrayList<>();
+
+    // If no trust chain path is provided, return an empty list.
+    if (Strings.isNullOrEmpty(trustChainPath)) {
+      return certificateTrustChain;
     }
+
+    // initialize certificate factory to retrieve x509 certificates.
+    CertificateFactory cf = CertificateFactory.getInstance("X.509");
+
+    // Read the trust chain file.
+    byte[] trustChainData;
+    trustChainData = Files.readAllBytes(Paths.get(trustChainPath));
+
+    // Split the file content into PEM certificate blocks.
+    String content = new String(trustChainData);
```

Comment:
I wonder if it makes sense to be a bit safer with charset specified?
```suggestion
    String content = new String(trustChainData, StandardCharsets.UTF_8);
```